rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,777 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 136 - Modern Signature Algorithms
+
+## Required Approvers
+
+* Engineering: (@jakule || @espadolini)
+* Security: (@reedloden || @jentfoo) && Doyensec
+* Product: @klizhentas
+
+## What
+
+This RFD proposes the adoption of a new set of asymmetric key types and
+signature algorithms across all Teleport protocols that surpass the 2048-bit RSA
+keys currently used in terms of security and performance.
+
+## Why
+
+Currently Teleport uses 2048-bit RSA keypairs for all internal Certificate
+Authorities and all user and host keys.
+
+More modern algorithms based on Elliptic Curve Cryptography offer better
+security properties with smaller keys that offer better performance for keypair
+generation, signing operations, and signature verification.  Some of the more
+restrictive security policies are starting to reject RSA2048 (e.g.
+[RHEL 8's FUTURE policy](https://access.redhat.com/articles/3642912)).
+
+## Details
+
+### Summary
+
+We will introduce the concept of signature algorithm "suites" that will
+configure the key types and signature algorithms used across Teleport.
+This will be a cluster-wide configuration available in both the `teleport.yaml`
+config of Teleport Auth servers (so that initial CA keys generated at first
+startup can be configured) and in the `cluster_auth_preference` (so that Cloud
+users can optionally change the suite used in their cluster).
+
+The initial set of suites we will introduce is:
+
+* `legacy`
+* `balanced-v1`
+* `fips-v1`
+* `hsm-v1`
+
+The `legacy` suite exists for compatibility and it will use the exact same
+key types and algorithms Teleport uses today (2048-bit RSA everywhere).
+This will be the initial default, changing the default to `balanced-v1` will
```

Comment:
to make this change, let's use our config versioning:

Existing config versions `v3` will all default to `legacy`. Introduce a `v4` config version that will switch to `balanced-v1` suite.